### PR TITLE
Rework of consumers during snapshot restore

### PR DESCRIPTION
**Pull request by derekcollison**
During rework of consumer restore during snapshot restore I reworked some raft internals and added forwarded proposals.
Lots of general cleanup and fixes for bugs and flappers. Trying to get stabilized on the testing front.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/raft.go [hunk 1]
```diff
@@ -1047,7 +1082,37 @@ func (n *raft) decodeAppendEntryResponse(msg []byte) *appendEntryResponse {
...
+		n.warn("Got error processing forwarded proposal: %v", err)
+	}
+}
+
 func (n *raft) runAsLeader() {
+	n.Lock()
+	// For forwarded proposals.
+	fsub, err := n.subscribe(n.psubj, n.handleForwardedProposal)
+	if err != nil {
+		n.warn("Error subscribing to forwarded proposals: %v", err)
```
Comment (kozlovic) [inline server/raft.go:1103]: Can clustering work without this subscription?

File: server/consumer.go [hunk 2]
```diff
@@ -525,15 +527,21 @@ func (mset *Stream) addConsumer(config *ConsumerConfig, oname string, node RaftN
...
 		// If we are here we have already registered this durable. If it is still active that is an error.
 		if eo.Active() {
+			o.name = _EMPTY_ // Prevent removal since same name.
+			o.deleteWithoutAdvisory()
 			return nil, fmt.Errorf("consumer already exists and is still active")
 		}
 		// Since we are here this means we have a potentially new durable so we should update here.
 		// Check that configs are the same.
 		if !configsEqualSansDelivery(o.config, eo.config) {
+			o.name = _EMPTY_ // Precent removal since same name.
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/consumer.go:543]: Will fix

File: server/consumer.go [hunk 3]
```diff
@@ -855,8 +875,62 @@ func (o *Consumer) updateDeliveryInterest(localInterest bool) {
...
+	acc, stream, name := o.acc.Name, o.stream, o.name
+	o.mu.RUnlock()
+
+	// If we are clustered, check if we still have this consumer assigned.
+	// If we do forward a proposal to delete ourselves to the metacontroller leader.
+	if s.JetStreamIsClustered() {
+		if ca := js.consumerAssignment(acc, stream, name); ca != nil {
+			// We copy and clear the reply since this removal is internal.
+			jsa.mu.Lock()
+			cca := *ca
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/consumer.go:898]: Good point and actually have the wrong lock anyway.. Will fix.

File: server/raft.go (hunk 1 above)
Comment (derekcollison) [reply server/raft.go:1103]: It can sans forwarded proposals. This should not happen ever, should I change to a panic?

Comment (derekcollison) [reply server/raft.go:1103]: Moved err check outside lock and made it a panic for now.

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**